### ISSUE TITLE
ddtrace/tracer: use sync.Pool for payload to improve performance

### ISSUE THREAD

**Issue by MichaelMure** [DataDog/dd-trace-go#948]
I noticed when using pprof on a production server that the datadog tracer is doing a non-trivial amount of heap memory allocation (about 13% in our case while the rest of the code is not especially memory efficient).

Zooming in and following the code, you can see that pretty much all of those allocations are done in [`payload.buf`](https://github.com/DataDog/dd-trace-go/blob/v1/ddtrace/tracer/payload.go#L45) when data accumulate there and the buffer need to grow:
![image](https://user-images.githubusercontent.com/294669/121173344-ffd8af00-c858-11eb-8b74-e5a544c11c6d.png)

Have you considered using `sync.Pool` to reuse those buffers instead of reallocating and growing the buffer for each trace? That would likely reduce the pressure on the GC and be more efficient.

**Comment by gbbr** [comment]
Hi Michael!

What I'm seeing in your graph is that `msgp.Encode` is using 100% of the time. I'm not sure I understand how a pool would improve that... Can you clarify?

**Comment by MichaelMure** [comment]
Yes, to be clear the flamegraph I posted show **the heap allocations during a time period**, not the memory used at a specific time (that is, not the "permanent" allocations). Unit is in bytes, not in number of allocations.

This view is also zoomed in on the tracer part (~13%) and ignore all the application specific part (87%), including some other small allocations when doing tracing (those are minor though). This is why it looks like 100%.

My point is, using a `sync.Pool` would almost entirely remove those 13% of allocations and reduce the pressure on the GC.

**Comment by gbbr** [comment]
Hi Michael!

I see what you mean. I haven't worked on this part for a while, but as far as I remember, we've tried this and had some issue with a race condition. The difficulty lies within the fact that the payload is sent to the transport [here](https://github.com/DataDog/dd-trace-go/blob/91b87dfa35b827bd8a8b1ce19acd683403bc4455/ddtrace/tracer/writer.go#L92), and then subsequently to the HTTP client  [here](https://github.com/DataDog/dd-trace-go/blob/91b87dfa35b827bd8a8b1ce19acd683403bc4455/ddtrace/tracer/transport.go#L164). This happens in [a go routine](https://github.com/DataDog/dd-trace-go/blob/91b87dfa35b827bd8a8b1ce19acd683403bc4455/ddtrace/tracer/writer.go#L84) while technically "in parallel" a [new payload](https://github.com/DataDog/dd-trace-go/blob/91b87dfa35b827bd8a8b1ce19acd683403bc4455/ddtrace/tracer/writer.go#L104) is prepared so that it can immediately be used while the flushing happens.

Now, if we were to place the "old" payload onto the pool after the HTTP send returns, it would present a risk because the HTTP client is known to potentially continue accessing the request body even after the `(*http.Client).Do` call returns. Granted, my memory of this event and attempt might be wrong. Also, since I've tried this last, we've added a [wait-close](https://github.com/DataDog/dd-trace-go/blob/91b87dfa35b827bd8a8b1ce19acd683403bc4455/ddtrace/tracer/transport.go#L168) mechanism to fix a similar race and that might ensure a payload is "done with".

Just some thoughts to keep in mind if we attempt this.

**Comment by knusbaum** [comment]
Yes, we've seen this issue from time to time. Reusing payload buffers is a great idea and would likely help the allocation issues.

I'm in support of investigating this and seeing if we can avoid the issues @gbbr mentioned.

**Comment by shannontan-bolt** [comment]
We're running into an odd but similar problem here where the majority of our heap consists of objects from `makeSlice` that aren't garbage collected. I'm quite lost as to how this has happened.

We're using Datadog's Continuous Profiler, and we can see that the makeSlice method somehow has 400MiB, which is about 80% of our heap live objects.

<img width="280" alt="Screen Shot 2021-07-06 at 5 31 45 PM" src="https://user-images.githubusercontent.com/45950434/124682652-2c93dc80-de80-11eb-92d1-8241b7fdf37f.png">

Situation is:
- We kick off a code path that spawns a lot of queries to our database, all of which are traced via Datadog's Gorm integration.
- We see spikes in our memory usage but then somehow the new steady state is significantly higher in memory usage and not being cleaned up by Golang's garbage collector.
- In the same time frame, we're seeing thousands of dropped traces from Datadog's Agent Trace Writer, and then we also see this from the Cluster agent: "2021-07-02 23:07:20 UTC | TRACE | WARN | (pkg/trace/writer/trace.go:252 in recordEvent) | Trace writer payload rejected by edge: 413 Request Entity Too Large"

Here's another screenshot of the profile (not the full profile):
<img width="576" alt="Screen Shot 2021-07-06 at 5 35 01 PM" src="https://user-images.githubusercontent.com/45950434/124682809-8a282900-de80-11eb-9de9-f730fbff6dab.png">

**Comment by shannontan-bolt** [comment]
As a result, we're having to turn off our tracer. We suspect that we ran into this problem now as compared to the past because Datadog advised us to use the tracing without limits and do ingestion control at the Datadog side. Let us know what you recommend or if there's something I'm missing.

**Comment by shannontan-bolt** [comment]
Also, to be clear, this profile is of "Heap Live Objects", not allocations.

**Comment by felixge** [comment]
@shannontan-bolt 👋 sorry to hear you're running into issues! I work on the profiler, but maybe I can help a little bit as well : ). 

We have an app we use for internal testing called go-prof-app that has an [endpoint](https://github.com/felixge/go-prof-app/blob/main/transaction_handler.go) that does 2 SQL queries and burns a bit of CPU time. It's instrumenting with profiling and tracing, and we're hitting it with 500 requests / second.

I just looked at the `Heap Live Size` for the app (using 1h aggregation), and the overhead I'm seeing for this code path (and the tracer in general) is much lower:

<img width="1701" alt="2021-07-07 Google Chrome Continuous Profiler | Datadog - 08-59-47" src="https://user-images.githubusercontent.com/15000/124714146-b20ba100-df01-11eb-8ffa-7d0aa1f5a87c.png">

The picture below shows all `Heap Live Size` allocations of the tracer highlighted in the overall picture.

<img width="1697" alt="2021-07-07 Google Chrome Continuous Profiler | Datadog - 09-01-31" src="https://user-images.githubusercontent.com/15000/124714339-ef702e80-df01-11eb-8699-4774a993fed9.png">

Based on this I suspect that the overhead you're seeing may not be inherent to tracing, and we might be able to figure out how to get rid of it.

My initial ideas on what might be going on:
- The agent is overloaded, forcing the ddtrace/tracer to hold on to payloads while retrying / waiting for confirmations. (not sure about the retries)
- The payloads are excessively large for some reason (huge values inside of the tags or something)

Would you be able to check if either of those two things might be happening in your environment? Seeing the graphs for [these tracer metrics](https://github.com/DataDog/dd-trace-go/blob/5f6503df3ce84003b8915cf999ff1a874e8039aa/ddtrace/tracer/writer.go#L88-L101) could also be interesting.

Again, I don't work on the tracer, so I might be wrong about some details here, but hopefully the above will be helpful.

**Comment by shannontan-bolt** [comment]
@felixge Thanks for the response. In terms of tracer metrics, we see above ~50k traces dropped. Flush duration spikes, but not considerably high (~160ms max 95percentile). Besides dropped traces + flush duration, is there anything else that we should look at?

"2021-07-02 23:07:20 UTC | TRACE | WARN | (pkg/trace/writer/trace.go:252 in recordEvent) | Trace writer payload rejected by edge: 413 Request Entity Too Large"

**Comment by gbbr** [comment]
@shannontan-bolt which agent version are you running? The agent part with the 413 response is possibly a separate issue.

**Comment by shannontan-bolt** [comment]
@gbbr looks like 7.24.1 (I think). I'm not a Kubernetes expert, but that looks like what we're pulling from GCR.

**Comment by shannontan-bolt** [comment]
@gbbr do you think agent version is related? I'm not entirely sure what is at fault here - it makes sense that the agent rejects too large of a payload, but that shouldn't lead to a buildup in the buffer that is unending.

**Comment by darccio** [comment]
@ahmed-mez Wouldn't #2319 solve this?

**Comment by radixdev** [comment]
Any motion on this? We're also seeing issues related to our heap growing almost entirely from Datadog traces.

**Comment by darccio** [comment]
@radixdev Is this a heap growth that looks like a memory leak or that tracing adds heap allocations? Our internal testing app doesn't show signs of any significant increase over time.

Have you used our Continuous Profiler to track down the allocations, as suggested in https://github.com/DataDog/dd-trace-go/issues/948#issuecomment-875352862? If you open a support ticket we could check it with you.

**Comment by radixdev** [comment]
I left our application running locally for a bit and just watched the allocation size grow to no end. 

<img width="2540" height="1273" alt="Image" src="https://github.com/user-attachments/assets/915c0684-9f3b-4928-b379-faef01a9e3a8" />

alloc_space is literally 68.54% from the tracer. https://gist.github.com/radixdev/6d1cb034ccb2fb8a7ec8df62a6541999

**Comment by radixdev** [comment]
Is there an issue with using a sync.Pool [here](https://github.com/DataDog/dd-trace-go/blob/886ce05e3005c3526bd00d6b6e115a293bdc8fc6/ddtrace/tracer/tracer.go#L704-L705) and returning the span in Finish [here](https://github.com/DataDog/dd-trace-go/blob/886ce05e3005c3526bd00d6b6e115a293bdc8fc6/ddtrace/tracer/span.go#L637-L638)?

**Comment by darccio** [comment]
Thanks for the additional information, @radixdev! We'll check it this week. There shouldn't be any issue to do this, but we want to double check it before getting into implementing it.

**Comment by radixdev** [comment]
Thanks @darccio . Also lmk if a PR from me would work as well. I can test out the changes on our app to do a red/green on the memory usage.

**Comment by darccio** [comment]
@radixdev We appreciate it but we want to provide it. It's in progress, and we'll share it with you so you can give it an early try once we deemed it as safe.

**Comment by radixdev** [comment]
Perfect! Thank you

**Comment by radixdev** [comment]
Any update on this? We're still very willing to demo this out once ready.

**Comment by darccio** [comment]
@radixdev It's a tricky thing to implement. As you can see, we did a second attempt to implement it, but we are pondering on what happens if a user still holds a reference to the span, as they usually do, after calling `Finish`.

**Comment by radixdev** [comment]
Not sure what your design philosophy in terms of API is, but I personally would be okay with your library just pushing that responsibility onto us as a client. As in, we'll just personally guarantee that we won't use the span after finish. If that also means that you only use the pool when we specifically enable that feature or enable it on a per-span basis, then I'd be okay with that as well.

**Comment by darccio** [comment]
@radixdev We understand, but right now that implementation is being worked as KTLO, not an OKR. This means it's developed as best effort and we want to be careful to avoid introducing behavioural breaking changes. Right now holding a span reference after `Finish()` is safe. This probably means that the span pool won't be enabled by default.

**Comment by darccio** [comment]
@radixdev @MichaelMure @sjmtan We are happy to tell you that we've introduced a span pool into `dd-trace-go`. It hasn't been released yet, but it'll be available as part of `v2.10.0` around end of July/start of August, as an experimental option `DD_TRACER_EXPERIMENTAL_SPAN_POOL_ENABLED`.

The semantics we settled for with the span pool is that anybody opting in must avoid interacting with the span reference after calling `Finish`. We are thinking about how to enable this in the long term in a way that can coexist with the invariant "holding a span reference after `Finish()` is safe". Probably some delayed pooling will be our best shot at this.

**Comment by radixdev** [comment]
Just to confirm, is this available now to use experimentally?
